File: stockflow-frontend/src/features/inventory/services/inventoryService.js
import api from '@/services/api';
import { API } from '@/constants/api';

/**
 * inventoryService — API service for Inventory & Stock Management endpoints.
 */
export const inventoryService = {
  stockIn: async (data) => {
    const res = await api.post(`${API.INVENTORY}/stock-in`, data);
    return res.data;
  },

  stockOut: async (data) => {
    const res = await api.post(`${API.INVENTORY}/stock-out`, data);
    return res.data;
  },

  adjustStock: async (data) => {
    const res = await api.post(`${API.INVENTORY}/adjust`, data);
    return res.data;
  },

  getProductInventory: async (productId) => {
    const res = await api.get(`${API.INVENTORY}/product/${productId}`);
    return res.data;
  },

  getProductHistory: async (productId, page = 0, size = 10, sortBy = 'movementDate', direction = 'DESC') => {
    const res = await api.get(`${API.INVENTORY}/history/${productId}`, {
      params: { page, size, sort: sortBy, direction },
    });
    return res.data;
  },

  getLowStockProducts: async (page = 0, size = 10, sortBy = 'quantity', direction = 'ASC') => {
    const res = await api.get(`${API.INVENTORY}/low-stock`, {
      params: { page, size, sort: sortBy, direction },
    });
    return res.data;
  },

  searchMovements: async (filters = {}, page = 0, size = 10, sortBy = 'movementDate', direction = 'DESC') => {
    const res = await api.get(`${API.INVENTORY}/movements/search`, {
      params: {
        productId: filters.productId || undefined,
        movementType: filters.movementType || undefined,
        dateFrom: filters.dateFrom || undefined,
        dateTo: filters.dateTo || undefined,
        username: filters.username || undefined,
        page,
        size,
        sort: sortBy,
        direction,
      },
    });
    return res.data;
  },
};

File: stockflow-backend/src/main/java/com/stockflow/backend/service/InventoryService.java
package com.stockflow.backend.service;

import com.stockflow.backend.dto.inventory.*;
import com.stockflow.backend.dto.product.ProductResponse;
import com.stockflow.backend.entity.MovementType;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDateTime;

/**
 * InventoryService — Core service interface for Inventory & Stock Management.
 */
public interface InventoryService {

    /**
     * Stock In — Increases product stock quantity and logs STOCK_IN movement.
     */
    StockMovementResponse stockIn(StockInRequest request, String username);

    /**
     * Stock Out — Decreases product stock quantity and logs STOCK_OUT movement.
     * Throws InsufficientStockException if requested > available.
     */
    StockMovementResponse stockOut(StockOutRequest request, String username);

    /**
     * Stock Adjustment — Sets product stock quantity to target absolute value and logs ADJUSTMENT movement.
     */
    StockMovementResponse adjustStock(StockAdjustmentRequest request, String username);

    /**
     * Get current inventory state for a product.
     */
    InventoryResponse getProductInventory(Long productId);

    /**
     * Get paginated stock movement history for a product.
     */
    Page<StockMovementResponse> getProductHistory(Long productId, Pageable pageable);

    /**
     * Get low stock products (quantity <= minimumStock).
     */
    Page<ProductResponse> getLowStockProducts(Pageable pageable);

    /**
     * Advanced search for stock movements.
     */
    Page<StockMovementResponse> searchMovements(
            Long productId,
            MovementType movementType,
            LocalDateTime dateFrom,
            LocalDateTime dateTo,
            String username,
            Pageable pageable);
}
